### PR TITLE
Allow fingerings in tablature staffs

### PR DESCRIPTION
**Pull request by MarkusLutz63**
As discussed in https://musescore.org/en/node/94976 it is necessary to have fingerings in lute tablature. With my pull request the possibility to select, wether you want fingerings (as in lute tablature, especially french tab) or not (as in guitar tablature), is added to the code, without influencing other behaviour.
The checkbox I have currently put in general style/score , but this could be discussed and changed.  
I have done the change in 3.0-devel, but it probably also should be included to 2.1 (what might need some changes ...).                                        
+ Allow fingerings in tablature staffs

### CODE REVIEW

**Comment by MarkusLutz63** [comment]
Probably it would be better to call it in the user interface "Show fingering in tablature staff". This is probably clearer and shorter!

File: libmscore/style.cpp [hunk 1]
```diff
@@ -543,6 +545,7 @@ static const StyleType styleTypes[] {
       { StyleIdx::tupletFontBold,                "tupletFontBold",               false },
       { StyleIdx::tupletFontItalic,              "tupletFontItalic",             true },
       { StyleIdx::tupletFontUnderline,           "tupletFontUnderline",          false },
+      { StyleIdx::tupletAlign,                   "tupletAlign",                  QVariant::fromValue(Align::CENTER) },
```
Comment (lasconic) [inline libmscore/style.cpp]: What is tupletAlign for? is not linked to this PR right?

Comment (MarkusLutz63) [inline libmscore/style.cpp]: Don't have to do anything with this PR indeed - don't know, what this means; at least I didn't include it deliberately ...

Comment (Jojo-Schmitz) [inline libmscore/style.cpp]: It is something from 94fe39f

File: libmscore/style.cpp [hunk 2]
```diff
@@ -937,6 +940,7 @@ const std::vector<StyledProperty> tupletStyle {
       { StyleIdx::tupletFontBold,                     P_ID::FONT_BOLD              },
       { StyleIdx::tupletFontItalic,                   P_ID::FONT_ITALIC            },
       { StyleIdx::tupletFontUnderline,                P_ID::FONT_UNDERLINE         },
+      { StyleIdx::tupletAlign,                        P_ID::ALIGN                  },
```
Comment (Jojo-Schmitz) [inline libmscore/style.cpp]: ditto

File: libmscore/style.h [hunk 3]
```diff
@@ -524,6 +525,7 @@ enum class StyleIdx {
       tupletFontBold,
       tupletFontItalic,
       tupletFontUnderline,
+      tupletAlign,
```
Comment (Jojo-Schmitz) [inline libmscore/style.h]: ditto

**Comment by lasconic** [comment]
For 2.1, we can't add style values. MuseScore 2.0 had a bug and will crash on any unknown style... 
As I said earlier on the mailing list I believe this option should be on the staff type level, probably in advanced properties for the tab staff type.

**Comment by MarcSabatella** [comment]
I agree.  You might want lute and guitar in the same score after all, with fingering for one but not the other.  Even without that possibility, it just makes more sense to make this a staff property.

**Comment by MarkusLutz63** [comment]
I'm not sure what this means for 2.1: Is there any possibility to add it to the staff property? Or is the bug also preventing that?
Another solution for 2.1 could be then, to allow fingerings for lute tablature and to reject it for guitar tablature (how can I best tell them from one another? Maybe testing on the sort of tablature, e.g. no fingerings in Spanish Tab?).
For 3.0 I would also prefer to add it to staff property, so that half of the PR probably has to be overworked, isn't it (the place of the variable and the ui especially).

**Comment by lasconic** [comment]
Staff properties should be fine for both 2.1 and 3.0. It needs to be tested for 2.1 though.
This PR needs to be reworked to move the property in staff type indeed and remove stuff that doesn't belong to it.

**Comment by MarkusLutz63** [comment]
How to do this as easy as possible? Can I again make the changes offline and reload them, or is there any other thing to do?

**Comment by lasconic** [comment]
You can keep on working on your local fork in the  176511-tablature-fingering branch. If you commit in this branch and then push to your origin (which should be your github fork at https://github.com/MarkusLutz63/MuseScore), it will automatically add the commits in this PR.

**Comment by lasconic** [comment]
If you need live help for that, feel free to join https://kiwiirc.com/client/irc.freenode.net/?nick=musescore|?#mlutz

File: Musescore.config [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// Add predefined macros for your project here. For example:
+// #define THE_ANSWER 42
```
Comment (Jojo-Schmitz) [inline Musescore.config]: This file should not be in a commit

File: Musescore.creator [hunk 5]
```diff
@@ -0,0 +1 @@
+[General]
```
Comment (Jojo-Schmitz) [inline Musescore.creator]: same here

File: Musescore.includes [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+thirdparty/poppler/splash
+thirdparty/portmidi/pm_common
+thirdparty/portmidi/pm_mac
+thirdparty/portmidi/pm_win
+thirdparty/portmidi/porttime
+thirdparty/qzip
+thirdparty/rtf2html
+thirdparty/singleapp/src
+thirdparty/xmlstream
+zerberus
```
Comment (Jojo-Schmitz) [inline Musescore.includes]: and here

File: libmscore/note.cpp [hunk 7]
```diff
@@ -558,8 +558,7 @@ Note::~Note()
 Note::Note(const Note& n, bool link)
    : Element(n)
       {
-    bool TabFingering = (score()->styleB(StyleIdx::TabFingeringYes));
-      if (link)
+       if (link)
```
Comment (Jojo-Schmitz) [inline libmscore/note.cpp]: wrong indent, we use 6 spaces, not 7

File: mscore/revision.h [hunk 8]
```diff
@@ -1 +1 @@
-3543170
+57d3883
```
Comment (Jojo-Schmitz) [inline mscore/revision.h]: this file shouldn't be in a commit. see https://musescore.org/en/developers-handbook/compilation/compile-instructions-windows-mingw-git#Source-Code, the part about post-checkout hook

File: Musescore.files [hunk 9]
```diff
@@ -0,0 +1,1948 @@
...
+zerberus/sfz.cpp
+zerberus/voice.cpp
+zerberus/voice.h
+zerberus/zerberus.cpp
+zerberus/zerberus.h
+zerberus/zerberusgui.cpp
+zerberus/zerberusgui.h
+zerberus/zone.cpp
+zerberus/zone.h
+all.h
```
Comment (Jojo-Schmitz) [inline Musescore.files]: and here

File: mscore/revision.h (hunk 8 above)
Comment (MarkusLutz63) [inline mscore/revision.h]: OK, I will change that - it isn't in https://musescore.org/en/developers-handbook/compilation/compile-instructions-linux-bsd-git ...

File: Musescore.files (hunk 9 above)
Comment (MarkusLutz63) [inline Musescore.files]: Does the above mentioned also help to exclude these files?

File: Musescore.config (hunk 4 above)
Comment (Jojo-Schmitz) [inline Musescore.config]: See also https://musescore.org/en/developers-handbook/compilation/compile-instructions-linux-bsd-git#Qt-Creator-IDE---method-2

File: Musescore.files (hunk 9 above)
Comment (Jojo-Schmitz) [inline Musescore.files]: No, I don't think so, but see https://musescore.org/en/developers-handbook/compilation/compile-instructions-linux-bsd-git#Qt-Creator-IDE---method-2

File: Musescore.config (hunk 4 above)
Comment (MarkusLutz63) [inline Musescore.config]: I copied that into the file .git/info/exclude - possibly that wasn't the correct place. No I saved it as "excludes".
How can I proceed to get the files away from the branch, does a new commit change that?

File: mscore/revision.h (hunk 8 above)
Comment (MarkusLutz63) [inline mscore/revision.h]: regarding undo.cpp: 
I think the code I commented out, should be there to not show fingerings in linked staves. Currently I cannot add any fingerings in such a tablature stave, if the option is not set, but they will be shown, if they are set in the notation system ... This shouldn't be at all, as this problem was reason for disabling fingerings in tablature lines.

**Comment by MarkusLutz63** [comment]
regarding undo.cpp:
I think the code I commented out, should be there to not show fingerings in linked staves. Currently I cannot add any fingerings in such a tablature stave, if the option is not set, but they will be shown, if they are set in the notation system ... This shouldn't be at all, as this problem was reason for disabling fingerings in tablature lines.

**Comment by Jojo-Schmitz** [comment]
There's a failure in mtest for tst_spanners, not sure this is related to this PR at all though

File: Musescore.config (hunk 4 above)
Comment (Jojo-Schmitz) [inline Musescore.config]: I'd do a `git reset 57d3883`, redo the commit without those files, then `git push -f`
Or even do the git rest to the commit before 77d3883, at the end of the day we'd want the commits squashed anyway

Comment (MarkusLutz63) [inline Musescore.config]: Will the changed files be safe, if I do a git reset?

Comment (MarkusLutz63) [inline Musescore.config]: Is this commit correct: 57d3883 , as below you give 77d3883 ? - I didn't find either of them in the master commits.

Comment (Jojo-Schmitz) [inline Musescore.config]: typo, sorry, I meant the one before 57d3883. And yes, your changes should be safe, as Long as you don't add a --hard

Comment (MarkusLutz63) [inline Musescore.config]: OK, I see, these are my commits ;-). Please be patient with me, as it is really not that easy to learn all these things ... 
Where do I find the number of the commit before 57d3883? Or will a naked git reset do that?

**Comment by Jojo-Schmitz** [comment]
Now there are 3 commits rather than just one?

File: libmscore/stafftype.cpp [hunk 10]
```diff
@@ -1,4 +1,4 @@
-//=============================================================================
+
```
Comment (Jojo-Schmitz) [inline libmscore/stafftype.cpp]: This changes doesn't seem needed (but doesn't harm either)

File: libmscore/note.cpp [hunk 11]
```diff
@@ -590,8 +590,9 @@ Note::Note(const Note& n, bool link)
             add(new Accidental(*(n._accidental)));
 
       // types in _el: SYMBOL, IMAGE, FINGERING, TEXT, BEND
+      bool TabFingering = staff()->staffType(tick())->showTabFingering();
```
Comment (lasconic) [inline libmscore/note.cpp]: please use `tabFingering` instead. We don't use capital for variables.

File: Musescore.config (hunk 4 above)
Comment (Jojo-Schmitz) [inline Musescore.config]: `git log` I think

File: libmscore/stafftype.cpp (hunk 10 above)
Comment (MarkusLutz63) [inline libmscore/stafftype.cpp]: This was probably a typo - sometimes my system (my laptop is about 7 years old) is hanging, or it seems at least. Probably the cursor had been in the wrong place ...

File: libmscore/note.cpp (hunk 11 above)
Comment (MarkusLutz63) [inline libmscore/note.cpp]: OK, I can change that.

**Comment by MarkusLutz63** [comment]
Should I do something regarding the many commits I have made? 
The problem was, that some of the corrections, I had made, had been missing after the reset - I reset on the 176511-tablature-fingering branch directly, as I didn't find any number before 57d3883.

BTW - the PR now gives the possibility to choose wether to use fingerings in tablature or not. You can choose this for every stave

**Comment by MarkusLutz63** [comment]
[](url)

**Comment by lasconic** [comment]
You can squash commit together. See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by MarkusLutz63** [comment]
Unfortunately I only could reduce it to two commits ...
I looked at the code to find the place, where I could control the showing of the fingerings: But I didn't find the place, where to do this.
Anyway I think, that the code now is more flexible, as you can choose, wether to use fingering or not to use. 
Do I have to do anything, if I want to have it in 2.1 also?

**Comment by MarkusLutz63** [comment]
Unfortunately I only could reduce it to two commits ...
I looked at the code to find the place, where I could control the showing of the fingerings: But I didn't find the place, where to do this.
Anyway I think, that the code now is more flexible, as you can choose, wether to use fingering or not to use. 
Do I have to do anything, if I want to have it in 2.1 also?

**Comment by Jojo-Schmitz** [comment]
Here's another way to get only one commit: first rebase, `git -pull --rebase musescore master`, then do a `git reset HEAD~2`, which should bring you to the last commit before your 2, then do the `git commit -m "Allow fingerings in tablature staffs"` followed by a `git push -f`.

**Comment by lasconic** [comment]
If you do the process highlighted by Jojo, please use "fix #176511: Allow fingerings in tablature staves" for your commit message. Having the issue number in it will help trace back the whole discussion in the future.

**Comment by Jojo-Schmitz** [comment]
Edited my comment accordingly

**Comment by lasconic** [comment]
For 2.1, once squashed and merged into master, I will try to merge it in 2.1 and let you know how it goes. If it goes well, we need to make sure that MuseScore 2.0.x can open the file without crashing and then document that this new feature will not be entirely compatible with previous versions.

**Comment by MarkusLutz63** [comment]
It's really frustrating: If I try to "git pull --rebase musescore master" (-pull was a typo, isn't it?), I get: fatal: 'musescore' does not appear to be a git repository
fatal: Could not read from remote repositor

"git reset HEAD~2" is also not working for me. I get the message: fatal: ambiguous argument 'HEAD~2': unknown revision or path not in the working tree. I tried this already earlier, when I merged the commits ... Should I try to do a pull from upstream, or will that mismatch everything?

**Comment by lasconic** [comment]
OK give me 10 minutes and I will try

**Comment by lasconic** [comment]
OK, it works for me. 
I checkout your branch then
```
$ git rebase master
$ git rebase -i HEAD~2
```
For the second line I replace `pick` by `s` and save the file
then I change the commit message to "fix #176511: Allow fingerings in tablature staves"  and save the file

I pushed the result on my fork and made a PR https://github.com/musescore/MuseScore/pull/3031

**Comment by Jojo-Schmitz** [comment]
yes the -pull``  was a typo. and instead of `musescore` it might be `upstream` for you.

**Comment by lasconic** [comment]
I'm waiting for the tests to pass on my PR and it should be good to merge. @MarkusLutz63 since you are the author of the commit, I need you to sign the MuseScore CLA at https://musescore.org/en/cla before merging.

**Pull request closed without merging**